tests/Collections/CachedConfigCollectionTest.php

```diff
@@ -56,6 +56,59 @@ public function testCacheHit()
         $collection->__destruct();
     }
 
+    public function testCacheOnDestruct()
+    {
+        $mockCache = $this->getMockBuilder(CacheInterface::class)->getMock();
+
+        $mockCollection = $this->getMockBuilder(ConfigCollectionInterface::class)->getMock();
+
+        // Cache will be generated, saved, and NOT saved again on __destruct()
+        $mockCache
+            ->expects($this->exactly(1))
+            ->method('set')
+            ->with(CachedConfigCollection::CACHE_KEY, $mockCollection);
+
+        $collection = new CachedConfigCollection();
+        $collection->setCollectionCreator(function () use ($mockCollection) {
+            return $mockCollection;
+        });
+        $collection->setCache($mockCache);
+
+        // Triggers cache set
+        $collection->getCollection();
+        // We have a cache, no more cache->set calls
+        $collection->getCollection();
+
+        // Do not write back changes to cache if there are no changes
+        $collection->__destruct();
+
+        $mockCache = $this->getMockBuilder(CacheInterface::class)->getMock();
```

Comment:
This part here is really confusing. At very least it should be split off into its own unit test.

I'm struggling to understand what it's actually testing though - seems that by mocking ConfigCollectionInterface we're simply overwriting CachedConfigCollection::get() ... and then testing that different stuff happened cos we overwrote a code path??  Doesn't seems like that's something we should be doing

I think perhaps we just delete from here downwards in this test unless I've missed something?

Reply:
@emteknetnz yes, it could be split in two, i just did it in the same test as a convenience. the goal here is to prove that when there are no update to the callCache, it's written only once (for example, when calling getCollection, but that's just one way to call the set method on the cache)
and then, in the second part of the test, that it still happens twice if there was an update to the callCache. 

maybe that's not necessary because it's already tested in the other tests, but at least i'm sure it fully covers that specific feature. I don't see any downside in having more test code that necessary, but i'm happy to remove the second half if you don't think it's necessary since it's already covered (indirectly) in the other tests

Reply:
OK, sounds like best thing then is to split it into a 2nd unit test